### PR TITLE
Removed YesAnd, CanAffordResponse, and PlayableCard through refactoring Think Tank, Reds Policy, and Card Warnings

### PR DESCRIPTION
**Pull request by EthanDobbs**

### CODE REVIEW

**Comment by EthanDobbs** [comment]
Some changes to a few of the card files are result of completely removing the 'tr' property, but then adding it back

File: src/server/cards/Card.ts [hunk 1]
```diff
@@ -222,30 +221,19 @@ export abstract class Card implements ICard {
   public get tilesBuilt(): ReadonlyArray<TileType> {
     return this.properties.tilesBuilt;
   }
-  public canPlay(player: IPlayer, canAffordOptions?: CanAffordOptions): boolean | YesAnd {
-    let yesAnd: YesAnd | undefined = undefined;
+
+  public canPlay(player: IPlayer, canAffordOptions?: CanAffordOptions): boolean {
```
Comment (kberg) [inline src/server/cards/Card.ts:225]: By the way do you have a better name for CanAffordOptions? If you do, I'd love to follow this up with a rename.

File: src/server/cards/ICard.ts [hunk 2]
```diff
@@ -120,24 +120,25 @@ export interface ICard {
 
   onProductionGain?(player: IPlayer, resource: Resource, amount: number): void;
   onProductionPhase?(player: IPlayer): void;
+  /** If a card has a mandatory effect that the player should be warned about (Pharmacy Union and Think Tank), the player's getWarning function will call this */
+  getWarningForCard?(player: IPlayer, card: ICard): string | undefined;
```
Comment (kberg) [inline src/server/cards/ICard.ts:124]: Why does ICard have a method that takes an ICard parameter?

File: src/server/cards/base/Flooding.ts [hunk 3]
```diff
@@ -17,7 +17,6 @@ export class Flooding extends Card implements IProjectCard {
       type: CardType.EVENT,
       name: CardName.FLOODING,
       cost: 7,
-      tr: {oceans: 1},
```
Comment (kberg) [inline src/server/cards/base/Flooding.ts:20]: Why is this removed?

File: src/server/cards/base/OlympusConference.ts [hunk 4]
```diff
@@ -57,7 +57,7 @@ export class OlympusConference extends Card implements IProjectCard {
         options.title = 'Select an option for Olympus Conference';
         return options;
       },
-      Priority.SUPERPOWER); // Unshift that deferred action
+      Priority.COST);
```
Comment (kberg) [inline src/server/cards/base/OlympusConference.ts:60]: Why is this no longer necessary?

File: src/server/cards/base/TerraformingGanymede.ts [hunk 5]
```diff
@@ -16,6 +15,10 @@ export class TerraformingGanymede extends Card implements IProjectCard {
       cost: 33,
       victoryPoints: 2,
 
+      behavior: {
```
Comment (kberg) [inline src/server/cards/base/TerraformingGanymede.ts:18]: Can you help me understand how this works correctly?

File: src/server/cards/pathfinders/SocialEvents.ts [hunk 6]
```diff
@@ -15,6 +14,10 @@ export class SocialEvents extends Card implements IProjectCard {
       cost: 18,
       tags: [Tag.EARTH, Tag.MARS],
 
+      behavior: {
```
Comment (kberg) [inline src/server/cards/pathfinders/SocialEvents.ts:17]: Same question viz: Terraforming Ganymede

File: src/server/cards/promo/PharmacyUnion.ts [hunk 7]
```diff
@@ -57,6 +57,13 @@ export class PharmacyUnion extends CorporationCard {
     return [Tag.MICROBE, Tag.MICROBE];
   }
 
+  public getWarningforCard(_player: IPlayer, card: ICard): string | undefined {
+    if (card.tags.includes(Tag.MICROBE)) {
+      return `Pharmacy Union loses 4M€ for every microbe tag played`;
```
Comment (kberg) [inline src/server/cards/promo/PharmacyUnion.ts:62]: Thank you. Let's use `common/cards/Warning.ts` for this.

File: src/server/cards/pathfinders/ThinkTank.ts [hunk 8]
```diff
@@ -35,6 +43,85 @@ export class ThinkTank extends ActionCard {
...
+  public onCardPlayed(player: IPlayer, card: ICard) {
+    player.game.defer(new SimpleDeferredAction(player, () => {
+      const dist = this.distance(card, player);
+      if (dist <= this.resourceCount) {
+        this.resourceCount -= dist;
+      } else {
+        throw new Error('Think Tank distance calculation error');
+      }
+      return undefined;
+    }, Priority.COST));
```
Comment (kberg) [inline src/server/cards/pathfinders/ThinkTank.ts:124]: This is not a cost behavior, so let's use something other than Priority.COST, which might apply to many of these.

File: src/server/cards/requirements/CardRequirement.ts [hunk 9]
```diff
@@ -55,5 +50,5 @@ export abstract class CardRequirement {
    * YesAnd if it can only do that under certain conditions (e.g. spend 2 Think Tank
```
Comment (kberg) [inline src/server/cards/requirements/CardRequirement.ts:50]: This comment can also go

File: tests/turmoil/Turmoil.spec.ts [hunk 10]
```diff
@@ -216,44 +216,47 @@ describe('Turmoil', function() {
...
+  /* it('Cannot play cards to raise TR via global parameters if Reds are ruling and player cannot pay', function() {
     setRulingParty(turmoil, game, new Reds());
     // Both of these cards cost 23MC.
     const iceAsteroid = new IceAsteroid();
     const protectedValley = new ProtectedValley();
     player.megaCredits = 25;
 
-    expect(player.canPlay(iceAsteroid)).is.not.true; // needs 29 MC
-    expect(player.canPlay(protectedValley)).is.not.true; // needs 26 MC
+    expect(player.canPlay(iceAsteroid)).is.false; // needs 29 MC
```
Comment (kberg) [inline tests/turmoil/Turmoil.spec.ts:238]: I appreciate the interest in making stylistic changes, but please avoid it. It not only complicates the review, but becomes another point of negotiation that I'd like to avoid for a while.

File: tests/turmoil/Turmoil.spec.ts [hunk 11]
```diff
@@ -216,44 +216,47 @@ describe('Turmoil', function() {
...
 
     player.production.add(Resource.ENERGY, 4);
     player.megaCredits = 30;
     const magneticFieldGeneratorsPromo = new MagneticFieldGeneratorsPromo();
-    expect(player.canPlay(magneticFieldGeneratorsPromo)).is.not.true; // needs 31 MC
+    expect(player.canPlay(magneticFieldGeneratorsPromo)).is.false; // needs 31 MC
   });
 
-  it('Cannot play cards to raise TR via global parameters if Reds are ruling and player cannot pay', function() {
+  // this test is fundementally flawed because adding oceans to the board gives the player placement bonuses
```
Comment (kberg) [inline tests/turmoil/Turmoil.spec.ts:228]: Use EmptyBoard, and/or just iterate through all the spaces and remove the bonuses. and you should be able  to restore the test.

**Review by kberg — COMMENTED** [review]
I've completed about half of the review,which helped me understand the first bit of it. I didn't look at Think Tank yet, which I'm very interested in dedicating time to.

In short, this looks like it will be a strong improvement. There are a few comments worth addressing for now, even before I review the rest of it.

File: src/server/cards/base/Flooding.ts (hunk 3 above)
Comment (EthanDobbs) [reply src/server/cards/base/Flooding.ts:20]: Remember how I originally wanted to remove the TR property? I never added it back for this card

File: src/server/cards/base/TerraformingGanymede.ts (hunk 5 above)
Comment (EthanDobbs) [reply src/server/cards/base/TerraformingGanymede.ts:18]: The function 'getTRSources' accounts for 'Countable' parameters with a counter object, and will recount the 'tr' behavior every time it's called.

File: src/server/cards/promo/PharmacyUnion.ts (hunk 7 above)
Comment (EthanDobbs) [reply src/server/cards/promo/PharmacyUnion.ts:62]: Will look into implementing that, but remember the function 'getWarningForCard()' returns a warning meant for the card given as a parameter to the function. So that would take wouldn't be a trivial change for me to make.

File: src/server/cards/ICard.ts (hunk 2 above)
Comment (EthanDobbs) [reply src/server/cards/ICard.ts:124]: I think this will be easier to explain with an example. If I want to play a card with a global req that would make me lose think tank resources, think tank should be responsible for giving that warning. The parameter would be 'Bushes' and the returned warning would be 'Playing bushes will remove 2 resources from think tank'.

File: src/server/cards/pathfinders/ThinkTank.ts (hunk 8 above)
Comment (EthanDobbs) [reply src/server/cards/pathfinders/ThinkTank.ts:124]: I would argue otherwise, but I see where you're coming from and will change this.

File: tests/turmoil/Turmoil.spec.ts (hunk 10 above)
Comment (EthanDobbs) [reply tests/turmoil/Turmoil.spec.ts:238]: Very understandable

File: tests/turmoil/Turmoil.spec.ts (hunk 11 above)
Comment (EthanDobbs) [reply tests/turmoil/Turmoil.spec.ts:228]: Would need to put in a more effort than that because oceans can still gain adjacency bonuses from each other. Will do this regardless.

File: src/server/cards/base/OlympusConference.ts (hunk 4 above)
Comment (EthanDobbs) [reply src/server/cards/base/OlympusConference.ts:60]: This was apart of removing Priority.SUPERPOWER all together, which honestly shouldn't have been packaged with this PR. Should just undo those changes here and open a new PR. The comments in the code said SUPERPOWER is a legacy thing and shouldn't be used anymore.

File: src/server/cards/base/Flooding.ts (hunk 3 above)
Comment (kberg) [reply src/server/cards/base/Flooding.ts:20]: Does that mean it should be added back?

File: src/server/cards/base/OlympusConference.ts (hunk 4 above)
Comment (kberg) [reply src/server/cards/base/OlympusConference.ts:60]: That's true. But this is a refactoring change, and this change is a behavioral change. let's keep behavioral changes separate from refactoring changes unless they're necessary. Addressing priorities is a separate thing, so let's remove this.

File: src/server/cards/promo/PharmacyUnion.ts (hunk 7 above)
Comment (kberg) [reply src/server/cards/promo/PharmacyUnion.ts:62]: Does card warnings have to be that big a change here?

File: tests/TestingUtils.ts [hunk 12]
```diff
@@ -136,7 +137,7 @@ export function formatMessage(message: Message | string): string {
  * @param passingDelta additional money required to take this action when Reds are in power.. Typically a multiple of 3
  * @param canAct set to true when cb calls canAct, which returns different results from canPlay. At the moment.
  */
-export function testRedsCosts(cb: () => CanPlayResponse, player: IPlayer, initialMegacredits: number, passingDelta: number, canAct: boolean = false) {
+export function testRedsCosts(cb: () => boolean, player: IPlayer, initialMegacredits: number, passingDelta: number, _canAct: boolean = false) {
```
Comment (kberg) [inline tests/TestingUtils.ts:140]: I think it's safe to remove _canAct entirely. This is nice.

File: tests/cards/pathfinders/Odyssey.spec.ts [hunk 13]
```diff
@@ -98,7 +98,7 @@ describe('Odyssey', () => {
     expect(player.playedCards).has.members([importOfAdvancedGHG, inventionContest]);
     expect(player.production.heat).eq(0);
 
-    selectProjectCardToPlay.payAndPlay(importOfAdvancedGHG, {...Payment.EMPTY, megaCredits: importOfAdvancedGHG.cost});
+    player.checkPaymentAndPlayCard(importOfAdvancedGHG, {...Payment.EMPTY, megaCredits: importOfAdvancedGHG.cost}, 'discard');
```
Comment (kberg) [inline tests/cards/pathfinders/Odyssey.spec.ts:101]: Please continue to use the card's action instead of deferring to player, e.g. selectProjectCardToPlay.cb();

File: tests/cards/pathfinders/Odyssey.spec.ts [hunk 14]
```diff
@@ -113,12 +113,11 @@ describe('Odyssey', () => {
     player.megaCredits = 50;
 
     player.playedCards = [importOfAdvancedGHG, mediaGroup];
-    const selectProjectCardToPlay = cast(odyssey.action(player), SelectProjectCardToPlay);
 
     expect(player.production.heat).eq(0);
     expect(player.megaCredits).eq(50);
 
-    selectProjectCardToPlay.payAndPlay(importOfAdvancedGHG, {...Payment.EMPTY, megaCredits: 9});
+    player.checkPaymentAndPlayCard(importOfAdvancedGHG, {...Payment.EMPTY, megaCredits: 9}, 'discard');
```
Comment (kberg) [inline tests/cards/pathfinders/Odyssey.spec.ts:120]: same

File: src/server/inputs/SelectProjectCardToPlay.ts [hunk 15]
```diff
@@ -93,20 +65,8 @@ export class SelectProjectCardToPlay extends BasePlayerInput<IProjectCard> {
     if (reserveUnits.plants + input.payment.plants > this.player.plants) {
       throw new Error(`${reserveUnits.titanium} units of plants must be reserved for ${input.card}`);
     }
-    const yesAnd = typeof(details.details) === 'boolean' ? undefined : details.details;
-    this.payAndPlay(card, input.payment, yesAnd);
-    return undefined;
-  }
-
-  public payAndPlay(card: IProjectCard, payment: Payment, yesAnd?: YesAnd) {
```
Comment (kberg) [inline src/server/inputs/SelectProjectCardToPlay.ts:101]: Thank you.

File: tests/cards/promo/ProjectInspection.spec.ts [hunk 16]
```diff
@@ -106,7 +106,7 @@ describe('ProjectInspection', function() {
     expect(action1.cards).has.lengthOf(2); // IW and PI (which was just played) are available
     expect(action1.cards[0]?.name).eq(indenturedWorkers.name);
     expect(action1.cards[1]?.name).eq(card.name);
-    action1.payAndPlay(card, Payment.EMPTY);
+    player.checkPaymentAndPlayCard(card, Payment.EMPTY, 'discard');
```
Comment (kberg) [inline tests/cards/promo/ProjectInspection.spec.ts:109]: Same. Stick with action.cb() or process().

action.cb() should work in all these cases. Let me know if they don't?

**Review by kberg — COMMENTED** [review]
Getting closer.

File: src/server/cards/base/OlympusConference.ts (hunk 4 above)
Comment (EthanDobbs) [reply src/server/cards/base/OlympusConference.ts:60]: Will do

File: tests/cards/promo/ProjectInspection.spec.ts (hunk 16 above)
Comment (EthanDobbs) [reply tests/cards/promo/ProjectInspection.spec.ts:109]: Will do. Also will add tests to make sure warnings for cards work properly

**Comment by kberg** [comment]
Feel free to reopen this and/or create a new PR when you are ready.
